crates/bindings/src/lib.rs

```diff
@@ -45,13 +42,610 @@ pub use timestamp::Timestamp;
+/// names.
+///
+/// ```ignore
+/// use spacetimedb::{BTreeIndex, UniqueColumn, Table, DbView};
+///
+/// // This generated struct is hidden and cannot be directly accessed.
+/// struct {name}__TableHandle { /* ... */ };
+///
+/// // It is a table handle.
+/// impl Table for {name}__TableHandle {
+///     type Row = {T};
+///     /* ... */
+/// }
+///
+/// // It can be looked up in a `ReducerContext`,
+/// // using `ctx.db().{name}()`.
+/// trait {name} {
+///     fn {name}(&self) -> Row = {T}>;
+/// }
+/// impl {name} for <ReducerContext as DbContext>::DbView { /* ... */ }
+///
+/// // Once looked up, it can be used to look up indexes.
+/// impl {name}Handle {
+///     // For each `#[unique]` or `#[primary_key]` field `{field}` of type `{F}`:
+///     fn {field}(&self) -> UniqueColumn<_, {F}, _> { /* ... */ };
+///     
+///     // For each named index `{index}` on fields of type `{(F1, ..., FN)}`:
+///     fn {index}(&self) -> BTreeIndex<_, {(F1, ..., FN)}, _>;
+/// }
+/// ```
+///
+/// [`Table<Row = Self>`]: `Table`
+#[doc(inline)]
+pub use spacetimedb_bindings_macro::table;
+
+/// Marks a function as a spacetimedb reducer.
+///
+/// A reducer is a function with read/write access to the database
+/// that can be invoked remotely by [clients].
+///
+/// Each reducer call runs in its own database transaction,
+/// and its updates to the database are only committed if the reducer returns successfully.
+///
+/// The first argument of a reducer is always a [`&ReducerContext`]. This context object
+/// allows accessing the database and viewing information about the caller, among other things.
+///
+/// After this, a reducer can take any number of arguments.
+/// These arguments must implement the [`SpacetimeType`], [`Serialize`], and [`Deserialize`] traits.
+/// All of these traits can be derived at once by marking a type with `#[derive(SpacetimeType)]`.
+///
+/// Reducers may return either `()` or `Result<(), E>` where `E: Debug`.
+///
+/// ```rust,ignore
+/// use spacetimedb::reducer;
+/// use log::info;
+///
+/// #[reducer]
+/// pub fn hello_world(context: &ReducerContext) {
+///     info!("Hello, World!");
+/// }
+///
+/// #[reducer]
+/// pub fn add_person(context: &ReducerContext, name: String, age: u16) {
+///     // add a "person" to the database.
+/// }
+///
+/// #[derive(SpacetimeType)]
+/// struct Coordinates {
+///     x: f32,
+///     y: f32,
+/// }
+///
+/// #[derive(Debug)]
+/// enum AddPlaceError {
+///     InvalidCoordinates(Coordinates),
+///     InvalidName(String),
+/// }
+///
+/// #[reducer]
+/// pub fn add_place(
+///     context: &ReducerContext,
+///     name: String,
+///     x: f32,
+///     y: f32,
+///     area: f32,
+/// ) -> Result<(), AddPlaceError> {
+///     // add a "place" to the database.
+/// }
+/// ```
+///
+/// Reducers may fail by returning a [`Result::Err`](std::result::Result) or by [panicking](std::panic!).
+/// Such a failure will be printed to the module logs and abort the active database transaction.
```

Comment:
I don't think we actually print `Err` returns to the module logs, only panic payloads. We may change that in the future, but I think we should be more vague here about how these messages actually get used as diagnostics. We should also mentioned that they get returned to the caller as a failure response.

Also, Tyler feels quite strongly that `panic` = bug, `Err` = expected failure, so you may want to mention that here.

Reply:
Not convinced that's a meaningful distinction in modules but whatever